Added a new StaticResourcesInfoProviderBase implementation that at runtime finds all registered Razor Pages

Stekeblad opened the pull request:
Fixes #20 and opens up many new possibilities for at runtime discovering resources to perform static generation on

Stekeblad commented on src/AspNetStatic/StaticGeneratorHostExtension.cs:
```diff
@@ -124,24 +115,39 @@ public static void GenerateStaticContent(
 						_httpClient.Timeout = TimeSpan.FromSeconds(httpTimeoutSeconds);
 						_httpClient.DefaultRequestHeaders.Add(HeaderNames.UserAgent, Consts.AspNetStatic);
 
+						// TODO: Something needs to change here. Create multiple configurations or move some options?
```
Any ideas on how to best support multiple resource providers here?

Stekeblad commented on src/AspNetStatic/StaticGeneratorHostExtension.cs:
Why is this class creating the HttpClient it is using the way it does? Can the named HttpClient configured in StaticGeneratorHostExtension be used here?

ZarehD commented:
@Stekeblad,

Sorry for the delay. Your algorithm is interesting but unfortunately this PR cannot be accepted. It makes too many unrelated and in my view unjustified changes.

The changes to `StaticGeneratorHostExtension` are not necessary and should be discarded. There's no need to request a collection of `IStaticResourcesInfoProvider` instances; you got rid of the empty-collection check on `IStaticResourcesInfoProvider`; and, it's not necessary to call `DiscoverResourcesAsync` from `GenerateStaticContent` - it can be called before the generator requests the `IStaticResourcesInfoProvider` instance.

I would also suggest that you derive `ActionDescriptorPageResourceInfoProvider ` (or its ancestor) from `StaticResourcesInfoProvider` in order to allow calling its methods on your type.

And a naming suggestion on your entry-point method; `DiscoverResourcesAsync` is somewhat ambigious in that it does not suggest what kind of resources it will attempt to discover. Perhaps you can rename it `DiscoverPageResourcesAsync`.

```csharp
...
public class ActionDescriptorPageResourceInfoProvider :
  StaticResourcesInfoProvider { ... }
...
builder.Services.AddSingleton<
  IStaticResourcesInfoProvider>(
  new ActionDescriptorPageResourceInfoProvider()
  .Add(builder.Environment.GetWebRootCssResources(["**/site.css", "**/*.min.css"]))
  .Add(builder.Environment.GetWebRootJsResources())
  .Add(builder.Environment.GetWebRootBinResources(["**/*.ico"]))
);
...
var p = 
  app.Services.GetRequiredService<IStaticResourcesInfoProvider>()
  as ActionDescriptorPageResourceInfoProvider;
p.DiscoverPageResourcesAsync(app.Services);
...
app.GenerateStaticContent(...);
```

Additionally, the code for this extension module is not confined enough. Indeed, upon further reflection, I think going forward it'd more appropriate for non-trivial extension implementations (such as this one) to be placed in separate projects alongside **AspNetStaticContrib** to allow users to pull in just the bits they wish to use. So, to that end, please...
- Move this implementation into a separate project (i.e. StekebladContrib.AspNetStatic).
- Add a `Readme.md` file in the new project describing how to use your contrib module.
- Build a NuGet package and push it to a public feed (i.e. nuget.org).

The changes to the sample project are also problematic in that they show using multiple instances of `IStaticResourcesInfoProvider` which is definitely not a necessity. It's also not necessary to change the sample project(s) to show an example usage of your entry point -- it's simple enough that you can present it in your module Readme, and if it's not, we can discuss creating a separate sample project for it.

And the remaining changes...
- Conts.cs: It's unrelated to your module, and I prefer to use `static readonly` to `const` here; so please discard these changes.
- Helpers.cs: No justification for removing the Uri well-formedness checks; please discard these changes.
- StaticGeneratorConfig.cs: Thank your for the `OfType<T>` suggestion; I will implement it directly so you can omit these changes.

Lastly, keep your pull request(s) targeted and focused and provide justification for the changes (preferably beforehand).

Thank you, @Stekeblad.

Stekeblad commented:
Sorry for my delay as well @ZarehD and thank you for taking your time to review and comment.

Your right about that I made some changes I did not have to, did not explain why anywhere or did not confirm some of them beforehand. I am new to open source contribution and got a bit lost in finding a working solution.

I can revert the unrelated changes of `const` and `OfType<T>`

I hope my answers below can answer many of the "why" questions you have about the choices I made.

---

> you got rid of the empty-collection check on `IStaticResourcesInfoProvider`

and

> I would also suggest that you derive `ActionDescriptorPageResourceInfoProvider`  (or its ancestor) from `StaticResourcesInfoProvider` in order to allow calling its methods on your type.


I removed the empty check of `PageResources` in `StaticGeneratorHostExtension` as a part of adding support for using multiple `IStaticResourcesInfoProvider`, it should be valid to have a resource provider only providing non-page resources.

 `StaticResourcesInfoProviderBase` exposes public methods for adding resources to the protected collection. I don't think you should be allowed to manually add resources to a provider that "dynamically" searches for resources. It adds extra complexity for provider implementations. I believe registering multiple providers is a better solution. Add static resources with a static provider and "dynamic" resources with a dynamic provider.

While I can't think about a good example right now; I believe being able to register multiple implementations of `DynamicResourceInfoProviderBase` would give a lot of power and possibilities to users of AspNetStatic.

> it's not necessary to call `DiscoverResourcesAsync` from `GenerateStaticContent` - it can be called before the generator requests the `IStaticResourcesInfoProvider` instance.

When do you mean it should be called then? Before adding the into the DI container? That is not possible as `DiscoverResourcesAsync` depends on the service provider being build and all services being registered. I added the call to `DiscoverResourcesAsync` inside `GenerateStaticContent` hoping that is the right place for it to discover new content if periodic regeneration of static content is active. (Imagine a custom provider fetching data from a CMS, it should be able to pick up on new content without restarting the application.)

> `DiscoverResourcesAsync` is somewhat ambigious in that it does not suggest what kind of resources it will attempt to discover.

`DiscoverResourcesAsync` is not named `DiscoverPageResourcesAsync` because it can be used to discover CssResources, JsResources and BinResources as well. I also think one discover method for everything is better than one for each type of resource. If a provider supports multiple types you could use the flags in `StaticResourcesInfoProviderBase` to exclude individual types of content or add your own properties to your custom provider.

> It's also not necessary to change the sample project(s) to show an example usage of your entry point -- it's simple enough that you can present it in your module Readme.

Your right. I should probably first showed an example of how to configure it in the PR description (or discussing it in the issue even earlier) and then add to readme/xmldoc later when refining before merge.

Regarding the Uri well-formedness: The project I have been testing my changes on have some paths containing non-english letters that incorrectly gets reported as invalid by `IsWellFormedUriString`. There are multiple issues in dotnet-repositories regarding this method and it's sibling `IsWellFormedOriginalString`. See for example this issue: https://github.com/dotnet/runtime/issues/72632
